fileconnector/fileconnector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/FileSearch.java

```diff
@@ -76,15 +82,29 @@ private void search(String source, String filePattern, MessageContext messageCon
                     OMElement result = factory.createOMElement(FileConstants.RESULT, ns);
                     resultPayload.preparePayload(messageContext, result);
                     FilePattenMatcher fpm = new FilePattenMatcher(filePattern);
+                    recursiveSearch = recursiveSearch.trim();
 
                     for (FileObject child : children) {
-                        if (child.getType() == FileType.FILE && fpm.validate(child.getName().
-                                getBaseName().toLowerCase())) {
-                            outputResult = child.getName().getBaseName();
-                            OMElement messageElement = factory.createOMElement(FileConstants.FILE,
-                                    ns);
-                            messageElement.setText(outputResult);
-                            result.addChild(messageElement);
+                        if (StringUtils.isEmpty(recursiveSearch) || recursiveSearch.equals("false")) {
+                            if (child.getType() == FileType.FILE && fpm.validate(child.getName().
+                                    getBaseName().toLowerCase())) {
+                                outputResult = child.getName().getPath();
+                                OMElement messageElement = factory.createOMElement(FileConstants.FILE,
+                                        ns);
+                                messageElement.setText(outputResult);
+                                result.addChild(messageElement);
+                            }
+                        } else if (recursiveSearch.equals("true")) {
+                            if (child.getType() == FileType.FILE && fpm.validate(child.getName().
+                                    getBaseName().toLowerCase())) {
+                                outputResult = child.getName().getBaseName();
+                                OMElement messageElement = factory.createOMElement(FileConstants.FILE,
+                                        ns);
+                                messageElement.setText(outputResult);
+                                result.addChild(messageElement);
+                            } else if (child.getType() == FileType.FOLDER) {
+                                searchSubFolders(child, filePattern, messageContext, factory, result, ns);
+                            }
                         }
```

Comment:
Avoid duplication of code. Following can be used to implement the above logic.

                      if (child.getType() == FileType.FILE && fpm.validate(child.getName().
                                     getBaseName().toLowerCase())) {
                                 outputResult = child.getName().getPath();
                                 OMElement messageElement = factory.createOMElement(FileConstants.FILE,
                                         ns);
                                 messageElement.setText(outputResult);
                                 result.addChild(messageElement);
                      } else if ( child.getType() == FileType.FOLDER && "false".equals(recursiveSearch)){
                                 searchSubFolders(child, filePattern, messageContext, factory, result, ns);
                      }